include/envoy/config/subscription.h

```diff
@@ -25,10 +25,109 @@ enum class ConfigUpdateFailureReason {
   UpdateRejected
 };
 
+/**
+ * A wrapper for xDS resources that have been deserialized from the wire.
+ */
+class DecodedResource {
+public:
+  virtual ~DecodedResource() = default;
+
+  /**
+   * @return const std::string& resource name.
+   */
+  virtual const std::string& name() const PURE;
+
+  /**
+   * @return const std::vector<std::string& resource aliases.
+   */
+  virtual const std::vector<std::string>& aliases() const PURE;
+
+  /**
+   * @return const std::string& resource version.
+   */
+  virtual const std::string& version() const PURE;
+
+  /**
+   * @return const Protobuf::Message& resource message reference. If hasResource() is false, this
+   *         will be the empty message.
+   */
+  virtual const Protobuf::Message& resource() const PURE;
+
+  /**
+   * @return bool does the xDS discovery response have a set resource payload?
+   */
+  virtual bool hasResource() const PURE;
```

Comment:
From an interface perspective you might consider just returning an optional reference from `resource()` and then I think you can do away with `hasResource()`?

Reply:
Yeah, I tried that initially, but it's cleaner to do it this way, since some sites still want a resource (the empty resource) even if `hasResource()` is false.